built-in-services/account/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+
+            if weight_sum >= permission.threshold {
+                return ServiceResponse::<VerifyResponse>::from_succeed(VerifyResponse {
+                    address: wit.sender,
+                });
+            }
+        }
+
+        ServiceResponse::<VerifyResponse>::from_error(
+            111,
+            "multi signature not verified".to_owned(),
+        )
+    }
+
+    #[cycles(100_00)]
+    #[read]
+    fn get_account_from_address(
+        &self,
+        ctx: ServiceContext,
+        payload: GetAccountPayload,
+    ) -> ServiceResponse<GenerateAccountResponse> {
+        let permission = self
+            .sdk
+            .get_account_value(&payload.user, &0u8)
+            .unwrap_or(Permission {
+                accounts:  Vec::<Account>::new(),
+                threshold: 0,
+            });
+
+        if permission.threshold == 0 {
+            return ServiceResponse::<GenerateAccountResponse>::from_error(
+                110,
+                "account not existed".to_owned(),
+            );
+        }
+
+        let mut accounts = Vec::<PayloadAccount>::new();
+        for item in &permission.accounts {
+            accounts.push(PayloadAccount {
+                address: item.address.clone(),
+                weight:  item.weight,
+            });
+        }
+
+        let response = GenerateAccountResponse {
+            accounts,
+            threshold: permission.threshold,
+            address: payload.user,
+        };
+
+        ServiceResponse::<GenerateAccountResponse>::from_succeed(response)
+    }
+
+    #[cycles(210_00)]
+    #[write]
+    fn generate_account(
+        &mut self,
+        ctx: ServiceContext,
+        payload: GenerateAccountPayload,
+    ) -> ServiceResponse<GenerateAccountResponse> {
+        if payload.accounts.is_empty() || payload.accounts.len() > MAX_PERMISSION_ACCOUNTS as usize
+        {
+            return ServiceResponse::<GenerateAccountResponse>::from_error(
+                110,
+                "accounts length must be [1,16]".to_owned(),
+            );
+        }
+
+        let mut weight_all = 0;
+        let mut accounts = Vec::<Account>::new();
+        for item in &payload.accounts {
+            weight_all += item.weight;
+            accounts.push(Account {
+                address:       item.address.clone(),
+                account_type:  ACCOUNT_TYPE_PUBLIC_KEY,
+                permission_id: 0,
+                weight:        item.weight,
+            });
+        }
+
+        if weight_all < payload.threshold || payload.threshold == 0 {
+            return ServiceResponse::<GenerateAccountResponse>::from_error(
+                110,
+                "accounts weight or threshold not valid".to_owned(),
+            );
+        }
+
+        let tx_hash = ctx.get_tx_hash().unwrap();
+
+        let keccak = HasherKeccak::new();
+        let addr_hash = Hash::from_bytes(Bytes::from(keccak.digest(&tx_hash.as_bytes())));
+        if addr_hash.is_err() {
```

Comment:
This is not a user error, and should never happen. You can just use `let addr_hash = Hash::from_bytes(Bytes::from(keccak.digest(&tx_hash.as_bytes()))).except("some meaningful msg");` to make it panic here.

Reply:
I am not sure in some case, e.g. tests , inner call or other way direct call ,maybe tx_hash is not valid

Reply:
Panic inside service will not crash the chain. The framework will catch the panic and stop the executing process instead of generating a transaction failure and writing it into the receipt. So it is better to panic for unexpected system error.

Reply:
ok